File: MCN/DL protocols/selective-repeat.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>

using namespace std;

int main() {
    int total_frames, window_size;
    cout << "Enter total frames: ";
    cin >> total_frames;
    cout << "Enter window size: ";
    cin >> window_size;

    srand(time(0));
    int acked[50] = {0}; 

    for (int i = 1; i <= total_frames; ) {
        for (int j = i; j < i + window_size && j <= total_frames; j++) {
            if (acked[j] == 0) {
                cout << "Sending Frame " << j << endl;
            }
        }

        for (int j = i; j < i + window_size && j <= total_frames; j++) {
            if (acked[j] == 0) {
                int ack = rand() % 4; 
                if (ack != 0) {
                    cout << "ACK received for Frame " << j << endl;
                    acked[j] = 1;
                } else {
                    cout << "ACK lost for Frame " << j << ". Will selectively repeat." << endl;
                }
            }
        }
        cout << endl;

        while (acked[i] == 1 && i <= total_frames) {
            i++;
        }
    }

    return 0;
}

File: MCN py/DL protocols/selective-repeat.py
import random

frames, window = map(int, input("Enter frames and window size: ").split())
acked = [False] * (frames + 1)
i = 1

while i <= frames:
    current_window = range(i, min(i + window, frames + 1))
    
    for j in current_window:
        if not acked[j]: print(f"Sending Frame {j}")
        
    for j in current_window:
        if not acked[j]:
            if random.random() > 0.2:
                print(f"ACK received for {j}")
                acked[j] = True
            else:
                print(f"ACK lost for {j}. Will repeat.")
                
    while i <= frames and acked[i]:
        i += 1
    print("-" * 20)